### ISSUE TITLE
surrogate gradient time must be finite and positive

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#636]
Training aborts with the error "surrogate gradient time must be finite and positive". The validation in `recipe.rs` (`number` helper) requires any numeric configuration value to be finite and greater than zero. The failure occurs before any training steps, indicating an invalid numeric parameter supplied to the recipe configuration (e.g., RECIPE_SURROGATE_RATE or another environment-provided number).

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `8347786a5db8bd57`
- Classifier provider: `agy`
- Classifier model: `gpt-oss-120b-medium`
- Classifier effort: `medium`

### Classifier decision

```json
{"body":"Training aborts with the error \"surrogate gradient time must be finite and positive\". The validation in `recipe.rs` (`number` helper) requires any numeric configuration value to be finite and greater than zero. The failure occurs before any training steps, indicating an invalid numeric parameter supplied to the recipe configuration (e.g., RECIPE_SURROGATE_RATE or another environment-provided number).","issue":0,"rationale":"Searches of the repository source and its GitHub issue tracker did not reveal any existing issue covering this validation failure. No prior issue addresses the \"must be finite and positive\" error in the training configuration path, so this packet represents a new public defect.","title":"surrogate gradient time must be finite and positive","verdict":"new"}
```

## Deterministic failure packet

id=8347786a5db8bd57
base=commit=10148e25421cc98c32cf7fbe59237263a1b98f73 tracked_tree=clean
cursor=seed:0 cursor:21186 next:21187 step:53196099145139 composition:124563731963814
data=path:"data/image/flat_sidecars" hash:cf6c8cf9aaeb42bc
configuration=data_mode:2 model:recipe.model().layer(4).sigmoid().iq(3).s.layer(4).prelu().norm(batch).iq(4).xs.loss(mse) loss:0 arithmetic:2 stop:2 rate:1 lifecycle:1
expected=training, optional resume, and inference produce finite numerical results through the public Recipe API
observed=phase:training message:surrogate gradient time must be finite and positive
output=phase:training message:surrogate gradient time must be finite and positive
replay=phase:training message:surrogate gradient time must be finite and positive stable:true
command=RECIPE_FORCE_CPU=1 cargo run --bin recipe -- /tmp/recipe-composition-repro.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = "/tmp/recipe-composition-repro.ogdl";
	let data = recipe.data("data/image/flat_sidecars")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.layer(4).sigmoid().iq(3).s
		.layer(4).prelu().norm(batch).iq(4).xs
		.loss(mse);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(124563731963814)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(32)
		.save(bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Does not reproduce on the unmodified base `minimal` at 1aa7e26f (`cargo build --release`, `recipe.rs` unchanged).

Command: the packet reproduction verbatim (only the bundle path moved out of `/tmp`), `RECIPE_FORCE_CPU=1 target/release/recipe repro.rs --device cpu`, fresh kernel cache per run, data `data/image/flat_sidecars`, seed 124563731963814, `.stop(0.8).fp(32).epochs(1)`.

- engi:cpu (Ryzen 5 7600X, 12 workers), md5 `7162ff51b72a3c3782b93512c8162602`: 9 of 9 runs exit 0, `run 1 mse 1.2461`, `eval 1 mse 1.2784`, `final_loss()` finite; no `surrogate gradient time must be finite and positive`.
- archy:cpu (Xeon E5-2620 v3, 12 workers, load 16-26), md5 `86201b57fab131a97c9274e6f4477025`: 4 of 4 runs exit 0, `measured archy:cpu 8.097264e7 work/s 0.002012259s/dispatch`, `run 1 mse 1.2461`, `eval 1 mse 1.2784`.

The message is not from the `number` configuration helper: it is `require(gradient > 0.0)` on `epoch - overhead` in `calibrate()`, the same abort as #585 (fingerprint 8347786a5db8bd57). The packet base 10148e25 timed one cold optimizer dispatch and one cold epoch each; the next commit on `minimal`, 4fc1259d "Warm both calibration dispatches before timing either", warms both paths and takes the minimum of 8 samples of each. On `minimal` the surrogate epoch measures 3-6x the dispatch (`overhead=0.000206735 epoch=0.000764428` on engi), so the difference stays positive.

**Comment by nm-z** [comment]
Packet id=8347786a5db8bd57 (same fingerprint 8347786a5db8bd57). Earliest failing public operation: recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).int(8).save(&bundle).run() in phase=training. Observed failure: panic at recipe.rs:21854 with message 'surrogate gradient time must be finite and positive'; replay phase=training stable:false. Implementation path matches #636 comment (calibrate() require(gradient > 0.0) on epoch - overhead, fingerprint 8347786a5db8bd57, same abort as #585). Different reproduction (data/numeric/compressed_csv.csv.gz, ensemble model with gru/rnn/layer) stops at the same earliest public operation with the same observed message. Existing #636 comments note non-reproduction on current source on comparable hosts (9/9 and 4/4 runs exit 0 with finite loss). No new source evidence changes classification.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `8347786a5db8bd57`
- Recipe source commit: `00bddd15f88ae7ce8b2674e4e1dd5edc48507ce1`
- Embedded Recipe `recipe.rs` blob: `e149eb47c95b7992c006a1a6a8f18616b5e07a85`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"comment","issue":636,"title":"surrogate gradient time must be finite and positive (training calibration failure)","body":"Packet id=8347786a5db8bd57 (same fingerprint 8347786a5db8bd57). Earliest failing public operation: recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).int(8).save(&bundle).run() in phase=training. Observed failure: panic at recipe.rs:21854 with message 'surrogate gradient time must be finite and positive'; replay phase=training stable:false. Implementation path matches #636 comment (calibrate() require(gradient > 0.0) on epoch - overhead, fingerprint 8347786a5db8bd57, same abort as #585). Different reproduction (data/numeric/compressed_csv.csv.gz, ensemble model with gru/rnn/layer) stops at the same earliest public operation with the same observed message. Existing #636 comments note non-reproduction on current source on comparable hosts (9/9 and 4/4 runs exit 0 with finite loss). No new source evidence changes classification.","rationale":"Same fingerprint and exact error message as #636. #636 comments explicitly identify this fingerprint as calibrate() positive-gradient check (not Config::load number() validation) and note it does not reproduce consistently on current source. The packet stops at the same earliest failing operation (recipe.train().run(), phase=training) with the same observed failure, so it is a duplicate measurement, not a new public defect."}
```

## Deterministic failure packet

id=8347786a5db8bd57
base=commit=00bddd15f88ae7ce8b2674e4e1dd5edc48507ce1 recipe_source=e149eb47c95b7992c006a1a6a8f18616b5e07a85
cursor=cursor:30474 next:30475 composition:30474
data=path:data/numeric/compressed_csv.csv.gz
configuration=cursor:30474 seed:17
expected=the generated public Recipe composition trains and infers with finite values
observed=phase:training message:surrogate gradient time must be finite and positive
output=phase:training
thread 'main' (1007822) panicked at recipe.rs:21854:13:
surrogate gradient time must be finite and positive
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
replay=phase:replay message:replay passed stable:false
command=RECIPE_DEVICE=aws-c7a target/debug/recipe test.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
    if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
        std::fs::write(path, value).expect("cannot write phase");
    }
}

fn width(path: &Path) -> usize {
    let text = std::fs::read_to_string(path).expect("cannot read saved model");
	let mut shape = text.lines().find_map(|line| line.trim().strip_prefix("shape ")).expect("saved model has no shape").split_whitespace();
    shape.next().expect("saved model has no channels").parse::<usize>().expect("saved channels are invalid") * shape.next().expect("saved model has no length").parse::<usize>().expect("saved length is invalid")
}

fn main() {
    let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
    let bundle = root.join(format!("recipe-model-{}.ogdl", std::process::id()));
    phase("setup");
    let data = recipe.data("data/numeric/compressed_csv.csv.gz").target("target");
    let model = recipe.model().svm().ensemble([gru(32).tanh().norm(l2), rnn(32).sigmoid().norm(batch).qi(5).q0, layer(32).qi(5).q0]).layer(1).loss(focal);
    phase("training");
    let report = recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).int(8).save(&bundle).run(&model, &data);
    assert!(report.final_loss().is_finite());
    phase("inference");
    let output = recipe.infer(&bundle, &vec![0.0; width(&bundle)]);
    assert!(!output.is_empty() && output.iter().all(|value| value.is_finite()));
    std::fs::remove_file(bundle).expect("cannot remove saved model");
}
```
backend=aws-c7a
